_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+I feel like I only just started learning good object-oriented design, and the world has moved on to functional programming! As a result, the OO principles that are fresh on my mind are exactly what people are moving away from. For example, as I've been checking out Elixir and Redux, I've had questions about whether a lack of certain OO features causes problems. For example, encapsulation allows you to isolate implementation knowledge to one part of your app, so that other parts of your app can use that functionality without being coupled to its implementation. Is it a problem that FP doesn't offer the same kind of encapsulation?
+
+It's difficult for me to find answers to these questions. As Yehuda Katz [tweeted][yehuda-1] [recently][yehuda-2], there's a tendency to describe programming advancements as radically new paradigms, and past approaches as terrible. There's no grappling with the arguments older approaches made.
+
+This leaves me with a conundrum: I bought into arguments about encapsulation increasing maintainability. Were those arguments true, or were they just something we made up to tell ourselves so we could build software in a way we liked?
+
+Even though these things aren't discussed much publicly, asking questions of people who know Elixir and Redux helped me with this.
+
+# Questioning Elixir and Redux
+
+First, I asked about the way data is separate from functionality in Elixir, and whether that caused problems. But I learned that when you define a struct inside a module, you also usually define functions that operate on that struct inside the same module. Because the name of the module and the name of the struct are the same, they go together naturally. So data and functionality aren't entirely separated.
+
+Then I asked about encapsulation: the data in a struct is available for everyone else in the app to access directly. But I learned that often you might have a convention of only accessing data in the struct via a function, and that that is a way to isolate knowledge of the data structure.
```

Comment:
They're not useless in communication, but I don't think it's a useful way to think about languages. It tends to fall apart when you look at it too closely.

Shriram Krishnamurthi actually has a great take on this as background for their textbook [Programming Languages: Application & Interpretation](http://www.plai.org) in the three-page article ["Teaching Programming Languages in a Post-Linnaean Age"](http://cs.brown.edu/~sk/Publications/Papers/Published/sk-teach-pl-post-linnaean/):

> Even more striking is what scripting languages say about the organization of languages. Most books rigorously adhere to the sacred division of languages into “functional”, “imperative”, “object-oriented”, and “logic” camps. I conjecture that this desire for taxonomy is an artifact of our science-envy from the early days of our discipline: a misguided attempt to follow the practice of science rather than its spirit.
>
> We are, however, a science of the artificial. What else to make of a language like Python, Ruby, or Perl? Their designers have no patience for the niceties of these Linnaean hierarchies; they borrow features as they wish, creating melanges that utterly defy characterization. How do we teach PL in this post-Linnaean era?

and later:

> If languages are not defined by taxonomies, how are they constructed? They are aggregations of features. Rather than study extant languages as a whole, which **conflates the essential with the accidental,** it is more instructive to decompose them into constituent features, which in turn can be studied individually. (emphasis added)

Going back to Harper's genomics vs cladistics: We still continue on with cladistics today, even in the face of genomics. We still use "zebra" to mean something in common language. But if you're going to spend time studying this, wrangling over an artificial taxonomy is probably less productive than investigating factual genomic details and grappling with what they might mean.